### PR TITLE
Explicitly close generators where the remote side closed early

### PR DESCRIPTION
**Pull request by jnwatson**
This prevents resources from hanging around indefinitely from a prematurely closed connection.

Resolves SYN-1734.

### CODE REVIEW

File: synapse/daemon.py [hunk 1]
```diff
@@ -181,6 +180,11 @@ async def t2call(link, meth, args, kwargs):
             else:
                 logger.exception('error during task %s', meth.__name__)
 
+            if isinstance(valu, types.AsyncGeneratorType):
```
Comment (vEpiphyte) [inline synapse/daemon.py:183]: Should this be done in a finally block?

Comment (invisig0th) [reply synapse/daemon.py:183]: i guess it depends on what happens if you call close on a closed generator

Comment (vEpiphyte) [reply synapse/daemon.py:183]: Just seems like we would want to be careful about when we are closing, in the event that closing a given generator throws (scary in and of itself); but closing them after we've sent the next lines feels safer for some reason. Multiple close calls should be safe.

Comment (jnwatson) [reply synapse/daemon.py:183]: There's no point is double closing a generator.

If a generator throws after being closed, that's a violation of the generator spec (and would normally be a RuntimeError).

I didn't want to close the link first as the closing generator might do something with it.

**Review by invisig0th — APPROVED** [review]

**Pull request merged**